lib/Fmt_ast.ml

```diff
@@ -3271,15 +3295,28 @@ and fmt_value_description ?ext c ctx vd =
     $ fmt_item_attributes c ~pre:(Break (1, 2)) atrs
     $ doc_after )
 
+and fmt_tydcl_param c ctx ty =
+  fmt_core_type ~tydecl_param:true c (sub_typ ~ctx ty)
+  $
+  match ty.ptyp_attributes with
+  | [] | _ :: _ :: _ -> noop
+  | [attr] -> fmt_if_k (is_layout attr) (fmt "@ :@ " $ str attr.attr_name.txt)
+
 and fmt_tydcl_params c ctx params =
-  fmt_if_k
-    (not (List.is_empty params))
-    ( wrap_fits_breaks_if ~space:false c.conf
-        (List.length params > 1)
-        "(" ")"
+  let empty, parenize =
+    match params with
+    | [] -> (true, false)
+    | [(p, _)] -> (
+        ( false
+        , match p.ptyp_attributes with
```

Comment:
Maybe abstract this out into a

```ocaml
val extract_layout_attr : attributes -> string option * attributes
```

which could be used in several places (and then also handle the case when more than one attribute is put on a type variable)?

Reply:
I know for some extensions we did the pattern you suggest of finding and filtering out extension attributes incrementally, but I think this is really an anti-pattern because of the subterm sanity checks in ocamlformat.

Many places in ocamlformat pass around a "context" containing the term you are in (here `ctx`) and do a sanity check that the term you are printing is an exact subterm of the context (the call to `sub_typ` below).  I think this system is meant to help catch bugs.

It's tempting to filter extension attributes from terms at the places where we have special logic for printing them, and pass an updated term to the normal printer for that syntactic form.  But generally that will cause us to fail these subterm checks because we don't have an exact subterm any more (the attribute is missing).  This makes ocamlformat fail loudly.

If you look at [the code that does this for local_ on arrows](https://github.com/janestreet/ocamlformat/blob/44f03ca6a551a48f4939f4996093a89820522a87/lib/Sugar.ml#L32), you'll see that to get around that we actually edit the context to remove the attribute there too.  But this is (a) hacky, (b) difficult to do in other places because the context doesn't always need to have just one shape like it does in that case, and (c) clearly going against the grain in terms of ocamlformat's design.

So I propose that even though it's not how we'd design things in a vacuum, the "right" think for adding our extensions to ocamlformat is to check for the presence of the extension attributes at the relevant places and handle them there, but don't remove them from the term, and then adjust the appropriate bits of attribute printing code to avoid printing them twice.  This is what Zesen did in his recent PR for `global_` in constructor arguments, for example.  It would be nice to go back and rework the places where we've done what I'm calling an antipattern, but it's not a priority.

All that said, I agree we'll want to pull out the logic for checking for the presence of a layout attribute.  But as described in the previous thread, the logic for type parameters in particular is a little weird due to limitations of the parser (which is why there is only a check in the case of a single attribute, here).  So it will make more sense to do that when we add v1.5 support and presumably have revisited the choices around type parameters.